stan/math/prim/mat/fun/dot_product.hpp

```diff
@@ -44,12 +44,32 @@ inline double dot_product(const double* v1, const double* v2, size_t length) {
  * @param v2 Second array.
  * @throw std::domain_error if the vectors are not the same size.
  */
-inline double dot_product(const std::vector<double>& v1,
-                          const std::vector<double>& v2) {
+inline double dot_product(const std::vector<double> &v1,
+                          const std::vector<double> &v2) {
   check_matching_sizes("dot_product", "v1", v1, "v2", v2);
   return dot_product(&v1[0], &v2[0], v1.size());
 }
-
+/**
+ * Returns the dot product of the specificied vector of elements
+ * @param x1 First vector
+ * @param x2 Second vector
+ * @param length size of vectors
+ * Assumes vectors are the same size
+ */
+template <typename T_1, typename T_2>
+inline typename stan::return_type<T_1, T_2>::type dot_product(
+    const std::vector<T_1> &x1, const std::vector<T_2> &x2, size_t length) {
+  T_1 result = 0;
+  for (size_t i = 0; i < length; ++i) {
+    result += x1[i] * x2[i];
```

Comment:
Rather than looping and adding, if the inputs are std::vectors you could use a single call to std::inner_product

Reply:
If this is templated, you can't call out to standard functions because they won't be implemented for our autodiff types.